lib/cache.js

```diff
@@ -4,7 +4,29 @@ const log = require('npmlog')
 const pacote = require('pacote')
 const path = require('path')
 const rimraf = promisify(require('rimraf'))
+const semver = require('semver')
 const BaseCommand = require('./base-command.js')
+const npa = require('npm-package-arg')
+
+const searchCachePackage = (spec, cacheKeys) => {
+  const parsed = npa(spec)
+  if (parsed.rawSpec !== '' && parsed.type === 'tag')
+    throw new Error(`Cannot list cache keys for a tagged package.`)
+  const searchMFH = new RegExp(`^make-fetch-happen:request-cache:.*(?<!/[@a-zA-Z]+)/${parsed.name}/-/([^/]+.tgz)$`)
```

Comment:
so this regex seems right for packages in the npm registry, but i want to raise a bit of a gotcha here so we can discuss it as part of the feature..

packuments are at a known path ahead of time, that's how we discover where tarballs are, so the `searchPack` regex below here is spot on. tarballs, however, are another story. there is no mandated path pattern for where a registry keeps its tarballs, we use the tarball url stored in the packument and fetch that. so while this feature will work for public registry users (or alternative registries that follow our path scheme), it is very likely it will miss tarball urls for third party registries.

_if_ we have a packument for the package in question, we could read its contents from the cache without sending a request and gather up the urls from there and look for those. if we don't have a packument, though, we're kinda stuck without sending a request..

Reply:
It now checks for keys from packument tarballs in addition to url parsing to cover registries that store packages at different paths. It still also checks for keys matching the public registry pattern.